src/core/Cargo.toml

```diff
@@ -0,0 +1,20 @@
+[package]
```

Comment:
Having these files live in `internal/core` would follow a pattern that's used for other subpackages containing native code, making this new code easier to learn. Is that technically straightforward enough to try?

Reply:
Easy enough to try.

I think the general pattern other repositories take is:

```
- package_name
  -  __init__.py
  - _native.so
- src
  - native.c
- pyproject.toml
```

https://github.com/pola-rs/polars/tree/main/py-polars

Just because others are doing it, doesn't mean we need to.

Eventually I'd like to see core stuff broken out into it's own package, rather than module.

Reply:
Something feels weird about it to me, but yes we can. Let me know if you'd prefer me to update it to nest under `internal/core/`

Reply:
FWIW, I tried the native-code-in-the-same-place-as-the-Python-code layout for [some recent changes to the Profiler](https://github.com/DataDog/dd-trace-py/tree/main/ddtrace/internal/datadog/profiling).  I actually don't love how it turned out, although I wasn't able to predict what would happen at the onset.  Knowing what I know now about this exact code, when I refactor this I'd break it up into a `src` pattern unless we have style guidance to the contrary.

Here are some thoughts.
1. Can't ever rely 100% on setup.py and associated fixtures like PyO3 or Pybind or whatever.  In my specific case, I have a native shared library I need to build and distribute (at load-time) to multiple extensions.  That means I need to carry detritus from a build system along.  Keeping the code together is really elegant and beautiful.  Keeping build cruft, scripts for the build, metadata, and other crap just makes it really hairy.
2. I like strong guarantees that I'm distributing only the native artifacts.  Over the course of extension development, I got basically every part of this wrong.  There's still some wrongness.  Having in-tree, non-distributable files is tricky.  In particular, I like _not_ having to tell the build system to exclude things.  In fact, I'd rather tell the build system precisely what I *do* want, since I presume failure to do so will be an error in testing.
3. I like .pyi quite a lot.  I think there's something to be said about specifying the *interface* to native code alongside the Python code, but putting the implementation itself somewhere else.  I think it sets an impractically high bar to assume everyone contributing to the repo is going to be able to grok the native-language-du-jour.  On the other hand, I think setting up explicit contracts at the interface (and documenting them in a .pyi) helps bridge this gap, while also helping us frame precisely what it is the native code is doing in terms of its interactions with Python.
4. Somewhat unrelated, but I've been radicalized to the idea that native components should be very careful in describing any off-GIL behavior they might bring.  In our current universe, we *will* be mixing forks and threads on Linux.  This is a horrible thing.


Opinions weakly held.  I'm only sharing my experiences.

Reply:
@emmettbutler what do you think?

Reply:
I'm convinced on `src`